connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDeserializer.scala

```diff
@@ -299,6 +298,29 @@ private[sql] class ProtobufDeserializer(
     }
   }
 
+  private def getFieldValue(record: DynamicMessage, field: FieldDescriptor): AnyRef = {
+    // We return a value if one of:
+    // - the field is repeated
+    // - the field is explicitly present in the serialized proto
+    // - the field is proto2 with a default
+    // - field presence is not available and materializeZeroValues is set
+    //
+    // Repeated fields have to be treated separately as they cannot have `hasField`
+    // called on them. And we only materialize zero values for fields without presence
+    // information because that flag controls the behavior for those fields in the ambiguous
+    // case of "unset" or "set to zero value". See the docs in [[ProtobufOptions]] for more
+    // details.
+    if (
+      field.isRepeated
+        || record.hasField(field)
+        || field.hasDefaultValue
+        || (!field.hasPresence && this.materializeZeroValues)) {
```

Comment:
I.e. if we replace `!field.hasPresence`  with `field.getJavaType != MESSAGE`, I think the tests pass.

Reply:
@rangadi I don't think this is the right approach. 

The goal of this PR is to give an option for developer who want Spark's protobuf to struct deserialization behavior comply to proto3 specs. As @justaparth mentioned earlier in [a discussion](https://github.com/apache/spark/pull/40686#discussion_r1160316173) with @viirya, the current behavior is something "make sense" (where unset/non-presense field set to null) but does not comply to proto3's spec (unset/non-presense field should be set to default value).

We think that Spark should provide an option for developer who need this behavior because this is the default behavior for many other protobuf library (especially those implemented in Go) so when user use Spark in conjunction with those libraries in different stages of their infra the behavior is consistent -- I believe there are ways to make the test pass, but that defeat the purpose.

Reply:
Well, I am not proposing change of behavior, but explicitly adding a check for Messages (through it is already covered in 'hasPresense()' check. At no point we want to serialize default messages. 

We should also explicitly mention in the documentation about messages. Most readers don't understands what 'has presence information'. They shouldn't have to read detailed protobuf spec to understand what this flag does. 

How about this: add the message check in addition to !hasPresense() check? (It's no op).

Reply:
In addition, note that this applies to Proto2 as well, not just Proto 3. This is correct. 

If you need to go strictly with "spec" we would need to materialize message types to since protobufs values are never null.

Reply:
> I.e. if we replace !field.hasPresence with field.getJavaType != MESSAGE, I think the tests pass.

actually, the tests, as written, will not pass. specifically, if we change `!field.hasPresence` to `field.getJavaType != message` then proto3 optional fields fields that were never set on the initial proto will get deserialized as their default value. the test [here](https://github.com/apache/spark/pull/40686/files#diff-f5dc81b1d20dde97f9d1dd26e993f811fa096bf68936ab5e449b9937314585d8R1196-R1239) checks that.

just to make sure, i did try this locally and verified the tests failed in the way described above.

> We should also explicitly mention in the documentation about messages. Most readers don't understands what 'has presence information' means. They shouldn't have to read detailed protobuf spec to understand what this flag does.
How about this: add the message check in addition to !hasPresense() check? (It's no op).

ah, if documentation / clarity is the issue i'm more than happy to write up some more comments or clear up the documentation!

Reply:
ah one thing here, the if statement is quite subtle as its written which i think is contributing to confusion :(. let me update the PR to make a small change here that I think will make this more clear

Reply:
okay! i've updated this if statement. I factored out the check for "should we materialize" into a function and added a good comment there! PTAL 🙏

Reply:
> lit(null).as("optional_int"), 

This [line at 1229](https://github.com/apache/spark/pull/40686/files#diff-f5dc81b1d20dde97f9d1dd26e993f811fa096bf68936ab5e449b9937314585d8R1229) ? 

I am more confused about the feature now. I thought all scaler fields like 'int' will have a value with this option set. 

What is the spec we are adhering to? What is the use case we are solving? May be we could update the flag documentation. 

Is this essentially 'materialize materialize protobuf V3 scalar fields that are not declared optional'? 

We are giving too much weightage to `optional` keyword it looks like. Logically all fields are optional in Proto3 by default. 

What happens to a `optional int` in V2 proto? Will it behave like proto3 `int` proto2 `optional int`?

Reply:
> Is this essentially materialize Protobuf V3 scalar fields that are not declared optional?


If this is the case, lets update the documentation clearly say this. Better not to say anything about 'presence information`. We shouldn't not require users to hunting down Protobuf documentation to understand it.

Plus, lets just add `field.getJavaType != message` in addition to `!hasPresense()` (though it is strictly not required. Better for readability.

Reply:
> Is this essentially materialize Protobuf V3 scalar fields that are not declared optional?

yes, this is the basic effect of this PR. this follows the proto3 spec https://protobuf.dev/programming-guides/proto3/#default. Maybe i could have made this more clear earlier

from the spec:
```
When a message is parsed, if the encoded message does not contain a particular singular element, the corresponding field in the parsed object is set to the default value for that field.
```

It's more precise to say that in cases where "field was never set" and "field was explicitly set to the zero value" are not distinguishable, (i.e. _field presence is not available_), the library deserializes both cases as  `null` by default, or to the type-specific zero value with this flag. This phrasing correctly covers every type in proto2 and proto3, and is consistent with the spec about field presence.

However, I understand your reservations that field presence is not the most accessible concept, and i'll update this documentation to say what you said above and then link to field presence documentation for people who want to read further.

> What happens to a optional int in V2 proto? Will it behave like proto3 int proto2 optional int?

no proto2 fields get default values in this implementation because theres no ambiguity.

Reply:
> If this is the case, lets update the documentation clearly say this. Better not to say anything about 'presence information`. We shouldn't not require users to go hunting down Protobuf documentation to understand it.

yeah, ill update it

Reply:
> Plus, lets just add field.getJavaType != message in addition to !hasPresense() (though it is strictly not required. Better for readability.

i feel like its a little bit odd to introduce code just for 1 specific type like this, especially given that it is redundant with the previous check. let's just leave a comment that indicates this behavior for future readers?

Reply:
I would prefer to have the check. Message type is pretty important. You have a test covering it, that is good. 
We don't want to serialize messages even if some future java library changes meaning of 'hasPresense' (though likely will not).

It is ok if you add a comment that it is really not required.

Reply:
> Message type is pretty important.
> You have a test covering it, that is good.

aren't all the types important? by this logic we should have a callout for every single type. i agree message is important and "not a scalar" so i thought a comment + test suite that explicitly asserts the behavior would help us prevent any problems

> We don't want to serialize messages even if some future java library changes meaning of 'hasPresence'

I don't think this would be possible, right? Because that would mean protobuf changes the sematics of their protocol, which doesn't seem possible (it would probably break proto parsing).

Reply:
but, in any case i've actually restructured the comment a bit, i wonder if its more clear now, do you want to take a look?

Reply:
> aren't all the types important? by this logic we should have a callout for every single type.

What? You want to exclude int etc? Message is different because it can contain other messages and easily blow up struct. Remember you could not explain if Messages would be serialized or not in earlier comment? Why, because it was not easy why they would be or wound not be included. You tested it, but didn't have an explanation. 

> The goal of this PR is to give an option for developer who want Spark's protobuf to struct deserialization behavior comply to proto3 specs. 

@pang-wu what spec is this? It will still have null for fields. I am ok with the feature, I don't understand the motivation.
We are providing any extra information in the Spark struct.
Could you given an example of problem this lets you solve?

Reply:
@rangadi the discussion of the spec behavior could be found [here](https://github.com/apache/spark/pull/40686#discussion_r1160291713)

Reply:
> Remember you could not explain if Messages would be serialized or not in earlier comment? Why, because it was not easy why they would be or wound not be included. You tested it, but didn't have an explanation.

If you check the thread where you originally asked:
https://github.com/apache/spark/pull/40686#discussion_r1179208375

I stated that message that weren't set in the original proto would remain null because message fields have presence information. I then linked a test asserting this exact behavior.

I don't think theres any confusion here?

> Message is different because it can contain other messages and easily blow up the struct. 

Yeah, i agree that message is an important type. This is why having a test to explicitly assert the deserialization, and a comment explaining it will cements the behavior. But i'm also trying to put myself in the shoes of a future reader of this code. If we add additional clauses to the conditional that are redundant, we are likely to confuse future readers who will inevitably ask "why is this code checking specifically for message types here", for which there isn't a fundamental answer, just that we had a desire in this PR to make sure message types were handled. But the tests / comments already should do that.